### PR TITLE
Add a "not in" operator to GDScript.

### PR DESCRIPTION
**Pull request by strank**
This provides the IMHO less controversial part of godotengine/godot-proposals#1319

(Outdated after patch revision: The implementation also requires a not-in operator in Variant which is actually
a small optimization of that conditional operator, as it avoids going through Variant again for the NOT-part.)

Here are also the diffs for my parser test-cases that show how parser behavior changed. I could add corresponding tests to the repo once godotengine/godot-proposals#1429 is in.

```
> cat tests/test_gdscript_doctest/not-is-not-an-operator.gd

var a = element not arr

--- a/tests/test_gdscript_doctest/not-is-not-an-operator.gd.parsed
+++ b/tests/test_gdscript_doctest/not-is-not-an-operator.gd.parsed
@@ -1,4 +1,4 @@
-02:17: Expected end of statement after variable declaration, found "not" instead.
+02:21: Expected "in" after "not" in content-test operator.
 Class <unnamed> :
 |   Variable a : Variant
-|   |   = element
+|   |   = (element NOT IN arr)
```

```
--- a/tests/test_gdscript_doctest/content-test.gd
+++ b/tests/test_gdscript_doctest/content-test.gd
@@ -1,4 +1,4 @@
 
 var a = element in arr
 var b = not element in arr
-#var c = element not in arr
+var c = element not in arr

--- a/tests/test_gdscript_doctest/content-test.gd.parsed
+++ b/tests/test_gdscript_doctest/content-test.gd.parsed
@@ -5,3 +5,5 @@ Class <unnamed> :
 |   Variable b : Variant
 |   |   = (NOT(element IN arr))
 
+|   Variable c : Variant
+|   |   = (NOT(element IN arr))
```

(Edit: updated the final test-case to reflect the revised patch. The `a not in b` is not a special operator anymore, so it gets parsed into exactly the same structure as `not (a in b)`.)

### CODE REVIEW

File: modules/gdscript/gdscript_parser.h [hunk 1]
```diff
@@ -344,6 +344,7 @@ class GDScriptParser {
 			OP_LOGIC_OR,
 			OP_TYPE_TEST,
 			OP_CONTENT_TEST,
+			OP_NOT_CONTENT_TEST,
```
Comment (vnen) [inline modules/gdscript/gdscript_parser.h:347]: I feel like this name is odd (it is not a content test, so what it is?). I would prefer something like `OP_NEGATIVE_CONTENT_TEST` or maybe `OP_ABSENCE_TEST`.

File: modules/visual_script/visual_script_nodes.cpp [hunk 2]
```diff
@@ -842,7 +842,8 @@ PropertyInfo VisualScriptOperator::get_input_value_port_info(int p_idx) const {
 		{ Variant::BOOL, Variant::BOOL }, //OP_XOR,
 		{ Variant::BOOL, Variant::BOOL }, //OP_NOT,
 		//containment
-		{ Variant::NIL, Variant::NIL } //OP_IN,
+		{ Variant::NIL, Variant::NIL }, //OP_IN,
```
Comment (vnen) [inline modules/visual_script/visual_script_nodes.cpp:845]: Not sure if this is really needed in Visual Script. In any case, without the operation in Variant this would need more work so it might be best to be done as another Pull Request.

**Review by vnen — CHANGES_REQUESTED** [review]

File: core/variant.h [hunk 3]
```diff
@@ -396,6 +396,7 @@ class Variant {
 		OP_NOT,
 		//containment
 		OP_IN,
+		OP_NOT_IN,
```
[the remark this answers is not in the document]
Comment (strank) [reply core/variant.h:399]: Sure, that works. I thought it would fit well with the other boolean operators (== and !=, > and <= and so on) and it would be a nicely shorter code-path during runtime for the price of adding the operator.

I'll change it to generate two operations.

File: modules/gdscript/gdscript_parser.cpp [hunk 4]
```diff
@@ -2025,6 +2025,16 @@ GDScriptParser::ExpressionNode *GDScriptParser::parse_unary_operator(ExpressionN
 	return operation;
 }
 
+GDScriptParser::ExpressionNode *GDScriptParser::parse_binary_not_in_operator(ExpressionNode *p_previous_operand, bool p_can_assign) {
+	// check that the NOT if followed by IN by consuming IN,
+	// but leave NOT in previous for the call to parse_binary_operator
+	// which can then treat it as shorthand for the "not in" infix operator:
+	GDScriptTokenizer::Token op = previous;
+	consume(GDScriptTokenizer::Token::IN, R"(Expected "in" after "not" in content-test operator.)");
+	previous = op;
```
[the remark this answers is not in the document]
Comment (strank) [reply modules/gdscript/gdscript_parser.cpp:2034]: I don't think it can be done inside the switch case - we have to consume the IN before parsing the right operand and that means checking for, and special-casing, `op == NOT` in every call to `parse_binary_operator`. Not a huge hit, performance-wise, but still easily avoidable by just adding this new function and switching to it in the prefix/infix table that will be checked anyway. It's the only operator that consists of two tokens, the new function seems to be a clean way to hide that from `parse_binary_operator`.

Changing `previous` was necessary since that indicates the type of operator inside `parse_binary_operator` and it's only one token, so NOT had to get there to differentiate it from a plain IN. But with the previous comment re: expanding NOT IN into to two operations instead of a new Variant operator, manipulating `previous` won't be necessary any more, so I'll change that.

However, I still think the new function is better than having an `if` in every call to `parse_binary_operator`. What do you think @vnen ?

File: modules/gdscript/gdscript_parser.cpp [hunk 5]
```diff
@@ -2637,7 +2651,7 @@ GDScriptParser::ParseRule *GDScriptParser::get_rule(GDScriptTokenizer::Token::Ty
 		// Logical
 		{ nullptr,                                          &GDScriptParser::parse_binary_operator,      	PREC_LOGIC_AND }, // AND,
 		{ nullptr,                                          &GDScriptParser::parse_binary_operator,      	PREC_LOGIC_OR }, // OR,
-		{ &GDScriptParser::parse_unary_operator,         	nullptr,                                        PREC_NONE }, // NOT,
+		{ &GDScriptParser::parse_unary_operator,         	&GDScriptParser::parse_binary_not_in_operator, PREC_CONTENT_TEST }, // NOT (unary) or NOT IN (binary),
```
[the remark this answers is not in the document]
Comment (strank) [reply modules/gdscript/gdscript_parser.cpp:2654]: Sure, I'll remove the more verbose note.

File: modules/gdscript/gdscript_parser.h (hunk 1 above)
Comment (strank) [reply modules/gdscript/gdscript_parser.h:347]: With the previous comment regarding not-adding-a-new-variant-op, this won't be needed anymore.
But just in case, we do decide to keep the Variant op: I like OP_NEGATIVE_CONTENT_TEST, more descriptive.

File: modules/visual_script/visual_script_nodes.cpp (hunk 2 above)
Comment (strank) [reply modules/visual_script/visual_script_nodes.cpp:845]: As above, not needed anymore, if we don't add a new op to Variant. Adding a NOT-IN Visual Script node is probably be a bit more complicated without the corresponding Variant op.

File: modules/gdscript/gdscript_parser.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (strank) [reply modules/gdscript/gdscript_parser.cpp]: One more note after updating the patch: using a separate function would also make it very easy to add other two-token operators such as "is not", the other part of godotengine/godot-proposals#1319. I would then rename the function to `parse_two_token_operator` for example, add it as the infix function for `is` as well, and have a simple switch or if/else that checks that `not` is followed by `is` or that `is` is optionally followed by `not`.

**Review by vnen — APPROVED** [review]

**Comment by vnen** [comment]
Completely forgot about this, sorry!

**Comment by akien-mga** [comment]
Thanks!
